### DISCUSSION TITLE
Can I rely on elements in glyph_info being sorted by cluster, and that the order matches the input order

### DISCUSSION THREAD

**Discussion by milasudril** [harfbuzz/harfbuzz#4948] [Q&A] [answered]
I need to find the begin and end of a selection, given a index pair of indices (the last index is excluded) into a std::u32string. For this, I use the following algorithm:

1. Search from the end of the glyph_info array, for a cluster number that is less than or equal to *begin*

2. Assuming there is a match i_start, search forward, starting at i_start + 1, for a value that is greater than or equal to *end* + 1. The result of this match is i_ind.

The algorithm

Assumes that the array of cluster numbers is ascending
Assumes that the order of clusters corresponds to the actual input order
Appears to work with ligatures (it cannot split the ligature though), in LTR mode.

Would my assumption break in case of BiDi or RTL order?

For reference: the  implementation is here: https://github.com/milasudril/terraformer/blob/1873d2fce113e912f9c7e22ef09ff330c2855d06/ui/font_handling/text_shaper.cpp#L100

**Comment by khaledhosny** [comment]
See https://harfbuzz.github.io/working-with-harfbuzz-clusters.html

**Reply by milasudril** [reply]
So RTL mode, it should search in the opposite direction?

**Reply by khaledhosny** [reply, marked as answer]
Yes.

**Reply by khaledhosny** [reply]
Make sure to read the whole clusters chapter in the user manual, it covers most of the clustering details: https://harfbuzz.github.io/clusters.html
